### PR TITLE
FEAT: add `bit_xor` aggregate to BigQuery and MySQL backends

### LINKED ISSUE #2641: FEAT: bit_xor aggregate
[BigQuery](https://cloud.google.com/bigquery/docs/reference/standard-sql/aggregate_functions#bit_xor), [MySQL](https://dev.mysql.com/doc/refman/5.7/en/aggregate-functions.html#function_bit-xor), [Spanner](https://cloud.google.com/spanner/docs/aggregate_functions?hl=da#bit_xor), [Sybase](http://infocenter.sybase.com/help/index.jsp?topic=/com.sybase.help.sqlanywhere.12.0.1/dbreference/bit-xor-function.html), and [Vertica](https://www.vertica.com/docs/9.2.x/HTML/Content/Authoring/SQLReferenceManual/Functions/Aggregate/BIT_XOR.htm) provide a `bit_xor` aggregate function, which can run a bitwise XOR operation on all 64-bit integers in a group (from a `GROUP BY` query).

I propose adding a `bit_xor` aggregate to integer values which compiles to the function of the same name.

### PR DESCRIPTION
**Pull request by tswast**
TODO:

- [x] Tests for BigQuery
- [ ] Tests for MySQL
- [x] Documentation

Closes #2641

### CODE REVIEW

**Comment by tswast** [comment]
I didn't see any examples of tests for the MySQL backend, so I'm omitting those for now.

File: ibis/expr/operations.py [hunk 1]
```diff
@@ -871,6 +871,14 @@ class Arbitrary(Reduction):
     output_type = rlz.scalar_like('arg')
 
 
+class BitXor(Reduction):
```
Comment (jreback) [inline ibis/expr/operations.py:874]: hmm do we have any other bit operations?

**Review by jreback — CHANGES_REQUESTED** [review]

File: ibis/expr/api.py [hunk 2]
```diff
@@ -1523,6 +1523,8 @@ def convert_base(arg, from_base, to_base):
 }
 
 
+bit_xor = _agg_function('bit_xor', ops.BitXor, True)
```
Comment (icexelloss) [inline ibis/expr/api.py:1526]: This should be potentially confusing as bit_xor can also be think of a elementwise operation between two columns. for example:

http://spark.apache.org/docs/latest/api/python/pyspark.sql.html?highlight=ntile#pyspark.sql.Column.bitwiseXOR

C++ also defines bit_xor as an operation that takes two value return compute bit_xor on them, which is similar to what bitwiseXOR in Spark.

Is there any API references that treats bit_xor as an aggregation function?

Comment (tswast) [reply ibis/expr/api.py:1526]: As mentioned in https://github.com/ibis-project/ibis/issues/2641 [BigQuery](https://cloud.google.com/bigquery/docs/reference/standard-sql/aggregate_functions#bit_xor), [MySQL](https://dev.mysql.com/doc/refman/5.7/en/aggregate-functions.html#function_bit-xor), [Spanner](https://cloud.google.com/spanner/docs/aggregate_functions?hl=da#bit_xor), and [Sybase](http://infocenter.sybase.com/help/index.jsp?topic=/com.sybase.help.sqlanywhere.12.0.1/dbreference/bit-xor-function.html) all provide a `bit_xor` aggregate function.

I'm okay with picking a different name to indicate that it's an aggregate, but went with `bit_xor` as it's the same in all of those systems.

File: ibis/expr/operations.py (hunk 1 above)
Comment (tswast) [reply ibis/expr/operations.py:874]: `bit_and` and `bit_or` are also available, though `bit_xor` is the only one suited for my checksum usecase. I could see `bit_or` being useful for a binary flags column, though.

File: ibis/expr/api.py (hunk 2 above)
Comment (tswast) [reply ibis/expr/api.py:1526]: Also called BIT_XOR in Veritca https://www.vertica.com/docs/9.2.x/HTML/Content/Authoring/SQLReferenceManual/Functions/Aggregate/BIT_XOR.htm

File: ibis/expr/operations.py (hunk 1 above)
Comment (tswast) [reply ibis/expr/operations.py:890]: I've pushed a commit adding `bit_and` and `bit_or` methods.

File: ibis/backends/bigquery/tests/test_compiler.py [hunk 3]
```diff
@@ -475,6 +475,60 @@ def test_approx_median(alltypes, project_id):
...
+SELECT BIT_XOR(`int_col`) AS `bit_xor`
+FROM `{project_id}.testing.functional_alltypes`"""
+    assert result == expected
+
+    b = alltypes.bigint_col
+    expr2 = i.bit_xor(where=b > 6)
+    result = expr2.compile()
+    expected = f"""\
+SELECT BIT_XOR(CASE WHEN `bigint_col` > 6 THEN `int_col` ELSE NULL END) AS `bit_xor`
+FROM `{project_id}.testing.functional_alltypes`"""  # noqa: E501
```
Comment (jreback) [inline ibis/backends/bigquery/tests/test_compiler.py:528]: what do these actually output? this is a xor operation per element?

File: ibis/expr/api.py [hunk 4]
```diff
@@ -1523,6 +1523,10 @@ def convert_base(arg, from_base, to_base):
 }
 
 
+bit_and = _agg_function('bit_and', ops.BitAnd, True)
```
Comment (jreback) [inline ibis/expr/api.py:1526]: can you add doc-strings to these, might be more clear what they do. Also add a link to the docs (bigquery / sql are fine)

File: ibis/backends/bigquery/tests/test_compiler.py (hunk 3 above)
Comment (tswast) [reply ibis/backends/bigquery/tests/test_compiler.py:528]: Yes? In a simile: As `SUM` is to `+`, `BIT_XOR` is to the traditional 2-argument `XOR`.

File: ibis/expr/api.py (hunk 4 above)
Comment (tswast) [reply ibis/expr/api.py:1527]: I updated the docstrings for the operators and pulled the logic from `_unary_op` that copies the docstring from the class into the `_agg_function` definition.

https://github.com/ibis-project/ibis/blob/f22e79f18449b99fb60552779ccaccd1aee74514/ibis/expr/api.py#L530

File: ibis/backends/bigquery/tests/test_compiler.py (hunk 3 above)
Comment (jreback) [reply ibis/backends/bigquery/tests/test_compiler.py:528]: i understand, can you show a hard coded example (in PR is fine), e.g. the actual output

**Comment by tswast** [comment]
Code example with these aggregations (exported from a test notebook):

```python
%load_ext google.cloud.bigquery
```


```python
%%bigquery
CREATE OR REPLACE TABLE my_dataset.re_flags_log
(
    flags INT64
)
```

    Query complete after 0.07s: 100%|██████████| 1/1 [00:00<00:00, 371.08query/s] 






```python
import re
```


```python
int(re.ASCII     | re.IGNORECASE)
```




    258




```python
int(re.LOCALE    | re.IGNORECASE)
```




    6




```python
int(re.LOCALE    | re.IGNORECASE | re.UNICODE)
```




    38




```python
int(re.MULTILINE | re.IGNORECASE)
```




    10



```python
int(re.VERBOSE   | re.IGNORECASE)
```




    66




```python
int(re.DOTALL    | re.IGNORECASE)
```




    18




```python
%%bigquery
INSERT INTO my_dataset.re_flags_log (flags) VALUES 
 (258), -- re.ASCII     | re.IGNORECASE
 (6),   -- re.LOCALE    | re.IGNORECASE
 (38),  -- re.LOCALE    | re.IGNORECASE | re.UNICODE
 (10),  -- re.MULTILINE | re.IGNORECASE
 (66),  -- re.VERBOSE   | re.IGNORECASE
 (18)   -- re.DOTALL    | re.IGNORECASE

```

    Query complete after 0.01s: 100%|██████████| 3/3 [00:00<00:00, 69.25query/s]                          






```python
import ibis.backends.bigquery
```


```python
bq = ibis.backends.bigquery.connect()
```


```python
flags_log = bq.database("my_dataset").table("re_flags_log")
```


```python
checksum = flags_log.flags.bit_xor()
print(checksum.compile())
checksum.execute()
```

    SELECT BIT_XOR(`flags`) AS `bit_xor`
    FROM `swast-scratch.my_dataset.re_flags_log`





    376




```python
all_flags = flags_log.flags.bit_or()
print(all_flags.compile())
all_flags.execute()
```

    SELECT BIT_OR(`flags`) AS `bit_or`
    FROM `swast-scratch.my_dataset.re_flags_log`





    382




```python
common_flags = flags_log.flags.bit_and()
print(common_flags.compile())
common_flags.execute()
```

    SELECT BIT_AND(`flags`) AS `bit_and`
    FROM `swast-scratch.my_dataset.re_flags_log`





    2


[pr2647.ipynb.txt](https://github.com/ibis-project/ibis/files/6071558/pr2647.ipynb.txt)

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @tswast 

if you can create an issue to implement this on other backends (pandas, others if possible). of course implementation PR would be great too :->

**Pull request merged**